### ISSUE TITLE
Circuits are not being cleaned up without traffic

### ISSUE THREAD

**Issue by Letalis82** [aspnet/AspNetCore#9893]
Hi,

Im writing a serverside blazor application where I keep the user session data in a Session scoped class. 
However I would like to dispose/clean the instance when it's no longer needed (when client disconnects or closes the page).

Any tips on how to detect this? Or should I implement an alive timer?

**Comment by BenHayat** [comment]
Great question;
I need something like that for my project. I suppose oe way to find out, would be through SignalR on the server side, when/if it loses connection with client.

**Comment by danroth27** [comment]
You can do this by implementing `IDisposable` on your root component and putting the handling logic in the `Dispose` method.

**Comment by Letalis82** [comment]
mm I can not see that happening. Looks like the disposed method is not called (at least not when I F5 or close the tab/browser).

I created a test solution:
https://github.com/Letalis82/BlazorDisposeExample

Class AppState is a Scoped class (and should be disposed when the user is disconnecting). 
The AppState is implementing IDisposable.
AppState is taged with a Guid (so that it can be identified). It starts a timer that notifies the Logger class (Singleton) that that id is still active. When AppState.Dispose is called the Logger is notified that this instance of AppState is disposed.

The FetchData page is showing the list of active and disposed AppState objects.

I tested with multiple browsers / tabs. I dont see disposes, F5 only triggers new AppStates to create a new instance.

Do I miss anything?

**Comment by danroth27** [comment]
I looked at this a bit with @pranavkm and it looks like we we are not correctly honoring the timeout for disposing circuits.

**Comment by mikewodarczyk** [comment]
It is not clear to me what the right way is to dispose() of the connection on the blazor client side.
The HubConnection methods that I want to use are async methods so in principle, I should be waiting for them somewhere with await or "GetAwaiter().GetResult()"

This code just hangs on the first Async step:

      public void Dispose()
        {
            Console.WriteLine("Try to leave the conversation");
            connection.InvokeAsync("LeaveTransferRequestId", TransferRequestId).GetAwaiter().GetResult();
            Console.WriteLine("Try to stop the connection");
            connection.StopAsync().GetAwaiter().GetResult();
            Console.WriteLine("call dispose on the connection");
            connection.Dispose();
            Console.WriteLine("Done");
        }

If I just dispose of the connection like this:

       public void Dispose()
        {
            Console.WriteLine("call dispose on the connection");
            connection.Dispose();
            Console.WriteLine("Done");
        }

Then I see in Chrome that the messages from the server keep coming to the connection after I leave the page.

This next code seems to work correctly:

      public void Dispose()
        {
            Console.WriteLine("Try to leave the conversation");
            connection.InvokeAsync("LeaveTransferRequestId", TransferRequestId);
            Console.WriteLine("Try to stop the connection");
            connection.StopAsync();
            Console.WriteLine("call dispose on the connection");
            connection.Dispose();
            Console.WriteLine("Done");
        }

but it seems like this is dangerous calling async methods and not waiting on them anywhere.  The server does seem to get the "LeaveTransferRequestId" method which takes my client out of the group and I don't get any messages after leaving the page.

It seems like we want/need an async event like "onBeforeUnloadAsync" or "onUnloadAsync" to clean up the SignalR connection the right way.

**Comment by pranavkm** [comment]
@javiercn picking this one up. Let me know if you have already started looking at this.
